internal/controllers/topology/cluster/structuredmerge/patchhelper.go

```diff
@@ -0,0 +1,132 @@
+		{"metadata", "namespace"},
+		// the topology controller controls/has an opinion for labels, annotation and spec only.
+		{"metadata", "labels"},
+		{"metadata", "annotations"},
+		{"spec"},
+	}
+
+	// If required, convert the original and modified objects to unstructured and filter out all the information
+	// not relevant for the topology controller.
+
+	var originalUnstructured *unstructured.Unstructured
+	if original != nil {
+		originalUnstructured = &unstructured.Unstructured{}
+		switch original.(type) {
+		case *unstructured.Unstructured:
+			originalUnstructured = original.DeepCopyObject().(*unstructured.Unstructured)
+		default:
+			if err := scheme.Convert(original, originalUnstructured, nil); err != nil {
+				return nil, errors.Wrap(err, "failed to convert original object to Unstructured")
+			}
+		}
+		filterObject(originalUnstructured, helperOptions)
+	}
+
+	modifiedUnstructured := &unstructured.Unstructured{}
+	switch modified.(type) {
+	case *unstructured.Unstructured:
+		modifiedUnstructured = modified.DeepCopyObject().(*unstructured.Unstructured)
+	default:
+		if err := scheme.Convert(modified, modifiedUnstructured, nil); err != nil {
+			return nil, errors.Wrap(err, "failed to convert original object to Unstructured")
+		}
+	}
+	filterObject(modifiedUnstructured, helperOptions)
+
+	// Determine if the intent defined in the modified object is going to trigger
+	// an actual change when running server side apply, and if this change might impact the object spec or not.
+	var hasChanges, hasSpecChanges bool
+	switch original {
+	case nil:
+		hasChanges, hasSpecChanges = true, true
+	default:
+		hasChanges, hasSpecChanges = dryRunPatch(&hasChangesContext{
+			path:     contract.Path{},
+			fieldsV1: getTopologyManagedFields(original),
+			original: originalUnstructured.Object,
+			modified: modifiedUnstructured.Object,
+		})
+	}
+
+	return &serverSidePatchHelper{
+		client:         c,
+		modified:       modifiedUnstructured,
+		hasChanges:     hasChanges,
+		hasSpecChanges: hasSpecChanges,
+	}, nil
+}
+
+// HasSpecChanges return true if the patch has changes to the spec field.
+func (h *serverSidePatchHelper) HasSpecChanges() bool {
+	return h.hasSpecChanges
+}
+
+// HasChanges return true if the patch has changes.
+func (h *serverSidePatchHelper) HasChanges() bool {
+	return h.hasChanges
+}
+
+// Patch will server side apply the current intent (the modified object.
+func (h *serverSidePatchHelper) Patch(ctx context.Context) error {
+	if !h.HasChanges() {
+		return nil
+	}
+
+	log := ctrl.LoggerFrom(ctx)
+	log.V(5).Info("Patching object", "Patch", string("TODO")) // TODO: we can eventually log the intent
+
+	options := []client.PatchOption{
+		client.FieldOwner(topologyManager),
+		// NOTE: we are using force ownership so in case of conflicts the topology controller
+		// overwrite values and become sole manager.
+		client.ForceOwnership,
+	}
+	return h.client.Patch(ctx, h.modified, client.Apply, options...)
```

Comment:
The tests in `cluster_controller_test.go` show that we also have to filter out the managedFields from the modifiedUnstructured object (at latest before we do the `h.client.Patch(..., client.Apply, ...)` here.

`h.client.Patch(..., client.Apply, ...)` requires objects which don't have managed fields set.